lib/web/resources.go

```diff
@@ -293,22 +305,58 @@ func upsertTrustedCluster(ctx context.Context, clt resourcesAPIGetter, content,
 	return ui.NewResourceItem(tc)
 }
 
-// CheckResourceUpsertableByError checks if the resource is upsertable by the state of error with
-// the request http method used.
-func CheckResourceUpsertableByError(err error, httpMethod, resourceName string) error {
-	if err != nil && !trace.IsNotFound(err) {
+// resourceExists returns a NotFound error if the resource does not exist.
+type resourceExists func(context.Context, string) error
+
+// CheckResourceUpsert checks if the resource can be created or updated, depending on the http method.
+func CheckResourceUpsert(ctx context.Context, httpMethod string, params httprouter.Params, payloadResourceName string, exists resourceExists) error {
+	switch httpMethod {
+	case http.MethodPost:
+		return trace.Wrap(checkResourceCreate(ctx, payloadResourceName, exists))
+	case http.MethodPut:
+		resourceName := params.ByName("name")
+		if resourceName == "" {
+			return trace.BadParameter("missing resource name")
+		}
+		return trace.Wrap(checkResourceUpdate(ctx, payloadResourceName, resourceName, exists))
+	default:
+		return trace.NotImplemented("http method %q not expected. this is a bug!", httpMethod)
+	}
+}
+
+// checkResourceCreate checks if the resource can be created, returning nil if it can.
+func checkResourceCreate(ctx context.Context, payloadResourceName string, exists resourceExists) error {
+	// Check if a resource with the name in the payload exists.
+	if err := exists(ctx, payloadResourceName); err != nil {
+		if trace.IsNotFound(err) {
+			// If the error is not found, then the resource does not exist and can be created.
+			return nil
+		}
 		return trace.Wrap(err)
 	}
 
-	exists := err == nil
-	if exists && httpMethod == http.MethodPost {
-		return trace.AlreadyExists("resource name %q already exists", resourceName)
+	// If no error, then the resource already exists and cannot be created.
+	return trace.AlreadyExists("resource name %q already exists", payloadResourceName)
```

Comment:
Just spitballing here, not sure what sounds best here. Perhaps even `resource %q already exists` would be enough but idk if that message makes it clear enough that you simply cannot have two resources with the same name.

```suggestion
	return trace.AlreadyExists("resource with name %q already exists", payloadResourceName)
```

Reply:
Done in 4e53ac1afc65dc7a306a43a6b376652114160844.